File: Samples/iOS/Objective-C/IOS-SDK-Demo-App/Ad Formats/LOBannerViewController.m
//
//  LOBannerViewController.m
//  test-objc-fw-integration
//
//  Created by Clarke Bishop on 3/29/23.
//

/**
 IMPORTANT: Please review and set the placementID inside the Info.plist file. To do so, please do the following:
 
 Navigate to project hierarchy in Xcode, and select Info. Expand Liftoff PlacementID and be sure that Banner is set correctly. Please do the same for Liftoff Dashboard AppID as well inside the Info.plist file.
 */

#import "LOBannerViewController.h"
#import <VungleAdsSDK/VungleAdsSDK.h>
#import "CallbackLog.h"
#import "AppUtil.h"
#import "CallbackLogTableViewCell.h"

@interface LOBannerViewController () <VungleBannerDelegate, UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) VungleBanner *bannerAd;
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, weak) IBOutlet UILabel *placementLbl;
@end

@implementation LOBannerViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  
  self.navigationItem.title = @"Liftoff Banner Ad";
  self.placementLbl.text = self.placementId;
  self.callbackLogs = [self generateCallbackLogs];
  
  self.tableView.delegate = self;
  self.tableView.dataSource = self;
  
}

#pragma mark - IBAction Methods

- (IBAction)btnPressed:(id)sender {
  if (sender == self.loadBtn && self.placementId != nil) {
    [self loadBannerAd];
  } else if (sender == self.playBtn && [self.bannerAd canPlayAd]) {
    [self playBannerAd];
  } else if (sender == self.closeBtn) {
    [self closeBannerAd];
  }
}

#pragma mark - Private Methods

- (void)loadBannerAd {
  // clean up before load
  if (self.bannerAd != nil) {
    self.bannerAd.delegate = nil;
    self.bannerAd = nil;
    [self resetCallbackLogs];
  }
  /**
   Banner Sizes:
      BannerSize.regular = 320x50
      BannerSize.short = 300x50
      BannerSize.leaderboard = 720x50
   */
  self.bannerAd = [[VungleBanner alloc] initWithPlacementId:self.placementId
                                                       size:BannerSizeRegular];
  self.bannerAd.delegate = self;
  [self.bannerAd load:nil];
}

- (void)playBannerAd {
  [self.bannerAd presentOn:self.bannerAdContainer];
}

- (void)closeBannerAd {
  for (id subview in self.bannerAdContainer.subviews) {
    [subview removeFromSuperview];
  }
}

- (NSArray *)generateCallbackLogs {
  return @[
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidLoad" index:0],
    [[CallbackLog alloc] initWithTitle:@"bannerAdWillPresent" index:1],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidPresent" index:2],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidTrackImpression" index:3],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidClick" index:4],
    [[CallbackLog alloc] initWithTitle:@"bannerAdWillLeaveApplication" index:5],
    [[CallbackLog alloc] initWithTitle:@"bannerAdWillClose" index:6],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidClose" index:7],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidFailToLoad" index:8],
    [[CallbackLog alloc] initWithTitle:@"bannerAdDidFailToPresent" index:9]
  ];
}

- (void)updateCallbackLogsWithTitle:(NSString *)title {
  dispatch_async(dispatch_get_main_queue(), ^{
    self.callbackLogs = [AppUtil updateCallbackLogs:self.callbackLogs title:title];
    [self.tableView reloadData];
  });
}

- (void)resetCallbackLogs {
  self.callbackLogs = [AppUtil resetCallbackLogs:self.callbackLogs];
  [self.tableView reloadData];
}

#pragma mark - UITableView Delegate / Datasource Methods

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.callbackLogs.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  CallbackLogTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CallbackLogTableCell" forIndexPath:indexPath];
  CallbackLog *log = self.callbackLogs[indexPath.row];
  if (log.isActive) {
    cell.callbackLbl.textColor = UIColor.labelColor;
    cell.accessoryType = UITableViewCellAccessoryCheckmark;
  } else {
    cell.callbackLbl.textColor = UIColor.grayColor;
    cell.accessoryType = UITableViewCellAccessoryNone;
  }
  cell.callbackLbl.text = log.title;
  return cell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  return 60;
}

#pragma mark - VungleBanner Delegate Methods

- (void)bannerAdDidLoad:(VungleBanner *)banner {
  NSLog(@"bannerAdDidLoad");
  [self updateCallbackLogsWithTitle:@"bannerAdDidLoad"];
}

- (void)bannerAdDidFailToLoad:(VungleBanner *)banner
                    withError:(NSError *)withError {
  NSLog(@"bannerAdDidFailToLoad");
  [self updateCallbackLogsWithTitle:@"bannerAdDidFailToLoad"];
}

- (void)bannerAdWillPresent:(VungleBanner *)banner {
  NSLog(@"bannerAdWillPresent");
  [self updateCallbackLogsWithTitle:@"bannerAdWillPresent"];
}

- (void)bannerAdDidPresent:(VungleBanner *)banner {
  NSLog(@"bannerAdDidPresent");
  [self updateCallbackLogsWithTitle:@"bannerAdDidPresent"];
}

- (void)bannerAdDidFailToPresent:(VungleBanner *)banner
                       withError:(NSError *)withError {
  NSLog(@"bannerAdDidFailToPresent");
  [self updateCallbackLogsWithTitle:@"bannerAdDidFailToPresent"];
}

- (void)bannerAdDidTrackImpression:(VungleBanner *)banner {
  NSLog(@"bannerAdDidTrackImpression");
  [self updateCallbackLogsWithTitle:@"bannerAdDidTrackImpression"];
}

- (void)bannerAdDidClick:(VungleBanner *)banner {
  NSLog(@"bannerAdDidClick");
  [self updateCallbackLogsWithTitle:@"bannerAdDidClick"];
}

- (void)bannerAdWillLeaveApplication:(VungleBanner *)banner {
  NSLog(@"bannerAdWillLeaveApplication");
  [self updateCallbackLogsWithTitle:@"bannerAdWillLeaveApplication"];
}

- (void)bannerAdWillClose:(VungleBanner *)banner {
  NSLog(@"bannerAdWillClose");
  [self updateCallbackLogsWithTitle:@"bannerAdWillClose"];
}

- (void)bannerAdDidClose:(VungleBanner *)banner {
  NSLog(@"bannerAdDidClose");
  [self updateCallbackLogsWithTitle:@"bannerAdDidClose"];
}

@end

File: Samples/iOS/Swift/IOS-SDK-Demo-App/Ad Formats/LOBannerViewController.swift
//
//  LOBannerViewController.swift
//  PublicSampleAppTrial
//
//  Created by John Mai on 12/29/22.
//

/**
 IMPORTANT: Please review and set the placementID inside the Info.plist file. To do so, please do the following:
 
 Navigate to project hierarchy in Xcode, and select Info. Expand Liftoff PlacementID and be sure that Banner is set correctly. Please do the same for Liftoff Dashboard AppID as well inside the Info.plist file.
 */

import UIKit
import VungleAdsSDK

class LOBannerViewController: UIViewController {
    @IBOutlet weak var bannerAdContainer: UIView!
    private var bannerAd: VungleBanner?
    
    @IBOutlet weak var loadBtn: UIButton!
    @IBOutlet weak var playBtn: UIButton!
    @IBOutlet weak var closeBtn: UIButton!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var placementLbl: UILabel!
    var callbackLogs: [CallbackLog] = []
    var placementId:String?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.title = "Liftoff Banner Ad"
        self.placementLbl.text = self.placementId
        self.callbackLogs = [
            CallbackLog(title: "bannerAdDidLoad", index: 0),
            CallbackLog(title: "bannerAdWillPresent", index: 1),
            CallbackLog(title: "bannerAdDidPresent", index: 2),
            CallbackLog(title: "bannerAdDidTrackImpression", index: 3),
            CallbackLog(title: "bannerAdDidClick", index: 4),
            CallbackLog(title: "bannerAdWillLeaveApplication", index: 5),
            CallbackLog(title: "bannerAdWillClose", index: 6),
            CallbackLog(title: "bannerAdDidClose", index: 7),
            CallbackLog(title: "bannerAdDidFailToLoad", index: 8),
            CallbackLog(title: "bannerAdDidFailToPresent", index: 9),
        ]
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
    }
    
    @IBAction func btnPressed(_ sender:UIButton) {
        if sender == loadBtn {
            if self.bannerAd != nil {
                self.bannerAd?.delegate = nil
                self.bannerAd = nil
                AppUtil.resetLogMessage(tableView: self.tableView, callbacks: self.callbackLogs)
            }
            guard let placementId = self.placementId else {
                return
            }
            
            /**
             Banner Sizes:
                BannerSize.regular = 320x50
                BannerSize.short = 300x50
                BannerSize.leaderboard = 720x50
             */
            self.bannerAd = VungleBanner(placementId: placementId, size: BannerSize.regular)
            self.bannerAd?.delegate = self
            self.bannerAd?.load()
        } else if sender == playBtn {
            // Determines whether the banner object is nil in addition to whether the ad assets have been downloaded successfully
            if ((self.bannerAd?.canPlayAd()) != nil) {
                self.bannerAd?.present(on: self.bannerAdContainer)
            }
        } else if sender == closeBtn {
            for subView in self.bannerAdContainer.subviews {
                subView.removeFromSuperview()
            }
        }
    }
}

// MARK: Vungle SDK Interstitial Delegate Callbacks

extension LOBannerViewController: VungleBannerDelegate {
    // Ad load Events
    func bannerAdDidLoad(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidLoad")
    }
    
    func bannerAdDidFailToLoad(_ banner: VungleBanner, withError: NSError) {
        print(#function)
        print(withError)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidFailToLoad")
    }
    
    // Ad Lifecycle Events
    func bannerAdWillPresent(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdWillPresent")
    }
    
    func bannerAdDidPresent(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidPresent")
    }
    
    func bannerAdDidFailToPresent(_ banner: VungleBanner, withError: NSError) {
        print(#function)
        print(withError)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidFailToPresent")
    }
    
    func bannerAdDidTrackImpression(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidTrackImpression")
    }
    
    func bannerAdDidClick(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidClick")
    }
    
    func bannerAdWillLeaveApplication(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdWillLeaveApplication")
    }
    
    func bannerAdWillClose(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdWillClose")
    }
    
    func bannerAdDidClose(_ banner: VungleBanner) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "bannerAdDidClose")
    }
}

extension LOBannerViewController:UITableViewDataSource, UITableViewDelegate {
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CallbackLogTableCell", for: indexPath) as! CallbackLogTableViewCell
        let log = self.callbackLogs[indexPath.row]
        if log.isActive {
            cell.callbackLbl.textColor = .label
            cell.accessoryType = .checkmark
        } else {
            cell.callbackLbl.textColor = .gray
            cell.accessoryType = .none
        }
        cell.callbackLbl?.text = log.title
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.callbackLogs.count
    }

}
